README.md

# pj

[![pub package](https://img.shields.io/pub/v/pj.svg?color=blue)](https://pub.dev/packages/pj)
<a href="https://github.com/cranst0n/pj/actions"><img src="https://github.com/cranst0n/pj/actions/workflows/main.yml/badge.svg" alt="Build Status"></a>
<a href="https://codecov.io/gh/cranst0n/pj"><img src="https://codecov.io/gh/cranst0n/pj/branch/main/graph/badge.svg" alt="codecov"></a>
<a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-green.svg" alt="License: MIT"></a>

Yet another JSON library build on top of [dartz](https://pub.dev/packages/dartz), loosely inspired by [circe](https://circe.github.io/circe/).

***Use at your own risk. Things are very much still in flux and I'm not concerned about breaking things yet.***

To get a quick sense of what this library provides, please look at the [example](https://github.com/cranst0n/pj/blob/main/example/main.dart). If it seems like something that may be useful to you, check out
the API reference for more combinators. 


CHANGELOG.md
# Changelog

## 0.0.10

- Make `Codec` constructor public.

## 0.0.9

- Bug fix for `Decoder.keyed`.
- **BREAKING CHANGE:** Renamed `keyed` to `at`.

## 0.0.8

- Add custom Encoder constructor.
- Bump minimum SDK to 2.16.0.

## 0.0.7

- Bug fix for Option Encoder.

## 0.0.6

- Increase arity to 20.

## 0.0.5

- Reworked the syntax API

## 0.0.4

- Add some missing primives.
- Various bug fixes and improvements.

## 0.0.3

- Add Codec API.
- Better naming consistency.

## 0.0.2

- Use ISO8061 format in DateTime encoder.

## 0.0.1

- Here be dragons


test/models.dart
import 'package:dartz/dartz.dart';
import 'package:equatable/equatable.dart';
import 'package:pj/pj.dart';
import 'package:pj/syntax.dart';

// ignore_for_file: avoid_positional_boolean_parameters

class Foo extends Equatable {
  final int a;
  final bool b;

  const Foo(this.a, this.b);

  static final codec = Codec.forProduct2(
    integer('a').withDefault(42),
    boolean('b'),
    Foo.new,
    (f) => tuple2(f.a, f.b),
  );

  @override
  List<Object?> get props => [a, b];
}

class Bar extends Equatable {
  final double a;
  final String b;

  const Bar(this.a, this.b);

  static final codec = Codec.forProduct2(
    dubble('a'),
    string('b'),
    Bar.new,
    (b) => tuple2(b.a, b.b),
  );

  @override
  List<Object?> get props => [a, b];
}

class Baz extends Equatable {
  final int integer;
  final Option<String> maybeString1;
  final Option<String> maybeString2;
  final double dubble;
  final bool boolean;
  final List<String> strings;
  final List<Foo> foos;
  final Bar bar;
  final List<bool> bools;
  final bool? nullable;
  final String recovered;
  final Map<String, dynamic> mary;
  final DateTime had;
  final Duration little;
  final BigInt lamb;

  const Baz(
    this.integer,
    this.maybeString1,
    this.maybeString2,
    this.dubble,
    this.boolean,
    this.strings,
    this.foos,
    this.bar,
    this.bools,
    this.nullable,
    this.recovered,
    this.mary,
    this.had,
    this.little,
    this.lamb,
  );

  static final codec = Codec.forProduct15(
    'integer'.integer.ensure((x) => x > 0, 'int must be > 0'),
    'maybeString1'.string.optional,
    'maybeString2'.string.optional,
    'dubble'.dubble,
    'boolean'.boolean,
    'strings'.listOf(Codec.string),
    'foos'.listOf(Foo.codec),
    'bar'.of(Bar.codec),
    'bools'.listOf(Codec.boolean),
    'nullable'.boolean.nullable,
    'recovered'.string.recover('recovered!'),
    'mary'.object,
    'had'.dateTime,
    'little'.duration,
    'lamb'.bigInt,
    Baz.new,
    (baz) => Tuple15(
      baz.integer,
      baz.maybeString1,
      baz.maybeString2,
      baz.dubble,
      baz.boolean,
      baz.strings,
      baz.foos,
      baz.bar,
      baz.bools,
      baz.nullable,
      baz.recovered,
      baz.mary,
      baz.had,
      baz.little,
      baz.lamb,
    ),
  );

  @override
  List<Object?> get props => [
        integer,
        maybeString1,
        maybeString2,
        dubble,
        boolean,
        strings,
        foos,
        bar,
        bools,
        nullable,
        recovered,
        mary,
        had,
        little,
        lamb,
      ];
}


test/codec_test.dart
import 'package:dartz/dartz.dart';
import 'package:pj/pj.dart';
import 'package:pj/syntax.dart';
import 'package:test/test.dart';

import 'arbitraries.dart';
import 'models.dart';

void main() {
  test('Codec.bigint', () {
    _codecTest(Codec.bigint, BigInt.one);
    _codecTest(Codec.bigint, BigInt.from(1234567890));
  });

  test('Codec.boolean', () {
    _codecTest(Codec.boolean, true);
    _codecTest(Codec.boolean, false);
  });

  test('Codec.dateTime', () {
    _codecTest(Codec.dateTime, DateTime.now());
    _codecTest(Codec.dateTime, DateTime.fromMillisecondsSinceEpoch(0));
  });

  test('Codec.dubble', () {
    _codecTest(Codec.dubble, 3.14);
    _codecTest(Codec.dubble, -8736.234);
  });

  test('Codec.duration', () {
    _codecTest(Codec.duration, Duration.zero);
    _codecTest(Codec.duration, const Duration(hours: 24));
  });

  test('Codec.xmap', () {
    final codec =
        Codec.string.xmap<bool>((s) => s.isNotEmpty, (b) => b ? 'xyz' : '');

    expect(codec.encode(false), '');
    expect(codec.encode(true), 'xyz');

    expect(codec.decode(''), right<DecodingError, bool>(false));
    expect(codec.decode('nonempty'), right<DecodingError, bool>(true));
  });

  test('Codec.exmap', () {
    Either<DecodingError, int> parseString(String s) =>
        optionOf(int.tryParse(s))
            .toEither(() => DecodingError('Failed to parse: $s'));

    final codec = string('foo').exmap<int>(parseString, (i) => i.toString());

    expect(codec.encode(42), {'foo': '42'});
    expect(codec.decode({'foo': '42'}), right<DecodingError, int>(42));
    expect(
        codec.decode({'foo': 'alpha'}),
        left<DecodingError, int>(
            const DecodingError('Failed to parse: alpha')));
  });

  test('Codec.recoverWith', () {
    final codec = Codec.integer.at('a').recoverWith(Codec.integer.at('b'));

    expect(codec.decode({'a': 1}), right<DecodingError, int>(1));
    expect(codec.decode({'b': 2}), right<DecodingError, int>(2));
    expect(codec.decode({'c': 3}),
        left<DecodingError, int>(DecodingError.missingField('b')));
  });

  test('Foo.codec', () {
    _codecTest(Foo.codec, Arbitraries.foo());
  });

  test('Codec.tuple2', () {
    _codecTest(
      Codec.tuple2(
        integer('a'),
        integer('b'),
      ),
      tuple2(1, 2),
    );
  });

  test('Codec.tuple3', () {
    _codecTest(
      Codec.tuple3(
        integer('a'),
        integer('b'),
        integer('c'),
      ),
      tuple3(1, 2, 3),
    );
  });

  test('Codec.tuple4', () {
    _codecTest(
      Codec.tuple4(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
      ),
      tuple4(1, 2, 3, 4),
    );
  });

  test('Codec.tuple5', () {
    _codecTest(
      Codec.tuple5(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
      ),
      tuple5(1, 2, 3, 4, 5),
    );
  });

  test('Codec.tuple6', () {
    _codecTest(
      Codec.tuple6(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
      ),
      tuple6(1, 2, 3, 4, 5, 6),
    );
  });

  test('Codec.tuple7', () {
    _codecTest(
      Codec.tuple7(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
      ),
      tuple7(1, 2, 3, 4, 5, 6, 7),
    );
  });

  test('Codec.tuple8', () {
    _codecTest(
      Codec.tuple8(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
      ),
      tuple8(1, 2, 3, 4, 5, 6, 7, 8),
    );
  });

  test('Codec.tuple9', () {
    _codecTest(
      Codec.tuple9(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
      ),
      tuple9(1, 2, 3, 4, 5, 6, 7, 8, 9),
    );
  });

  test('Codec.tuple10', () {
    _codecTest(
      Codec.tuple10(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
      ),
      tuple10(1, 2, 3, 4, 5, 6, 7, 8, 9, 10),
    );
  });

  test('Codec.tuple11', () {
    _codecTest(
      Codec.tuple11(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
      ),
      tuple11(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11),
    );
  });

  test('Codec.tuple12', () {
    _codecTest(
      Codec.tuple12(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
      ),
      tuple12(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12),
    );
  });

  test('Codec.tuple13', () {
    _codecTest(
      Codec.tuple13(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
      ),
      tuple13(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13),
    );
  });

  test('Codec.tuple14', () {
    _codecTest(
      Codec.tuple14(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
        integer('n'),
      ),
      tuple14(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14),
    );
  });

  test('Codec.tuple15', () {
    _codecTest(
      Codec.tuple15(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
        integer('n'),
        integer('o'),
      ),
      tuple15(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15),
    );
  });

  test('Codec.tuple16', () {
    _codecTest(
      Codec.tuple16(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
        integer('n'),
        integer('o'),
        integer('p'),
      ),
      tuple16(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16),
    );
  });

  test('Codec.tuple17', () {
    _codecTest(
      Codec.tuple17(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
        integer('n'),
        integer('o'),
        integer('p'),
        integer('q'),
      ),
      tuple17(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17),
    );
  });

  test('Codec.tuple18', () {
    _codecTest(
      Codec.tuple18(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
        integer('n'),
        integer('o'),
        integer('p'),
        integer('q'),
        integer('r'),
      ),
      tuple18(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18),
    );
  });

  test('Codec.tuple19', () {
    _codecTest(
      Codec.tuple19(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
        integer('n'),
        integer('o'),
        integer('p'),
        integer('q'),
        integer('r'),
        integer('s'),
      ),
      tuple19(
          1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19),
    );
  });

  test('Codec.tuple20', () {
    _codecTest(
      Codec.tuple20(
        integer('a'),
        integer('b'),
        integer('c'),
        integer('d'),
        integer('e'),
        integer('f'),
        integer('g'),
        integer('h'),
        integer('i'),
        integer('j'),
        integer('k'),
        integer('l'),
        integer('m'),
        integer('n'),
        integer('o'),
        integer('p'),
        integer('q'),
        integer('r'),
        integer('s'),
        integer('t'),
      ),
      tuple20(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
          20),
    );
  });

  test('Codec.forProduct2', () {
    _codecTest(
      Codec.forProduct2(
        'a'.integer,
        'b'.integer,
        (a, b) => tuple2(a, b),
        id,
      ),
      tuple2(1, 2),
    );
  });

  test('Codec.forProduct3', () {
    _codecTest(
      Codec.forProduct3(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        (a, b, c) => tuple3(a, b, c),
        id,
      ),
      tuple3(1, 2, 3),
    );
  });

  test('Codec.forProduct4', () {
    _codecTest(
      Codec.forProduct4(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        (a, b, c, d) => tuple4(a, b, c, d),
        id,
      ),
      tuple4(1, 2, 3, 4),
    );
  });

  test('Codec.forProduct5', () {
    _codecTest(
      Codec.forProduct5(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        (a, b, c, d, e) => tuple5(a, b, c, d, e),
        id,
      ),
      tuple5(1, 2, 3, 4, 5),
    );
  });

  test('Codec.forProduct6', () {
    _codecTest(
      Codec.forProduct6(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        (a, b, c, d, e, f) => tuple6(a, b, c, d, e, f),
        id,
      ),
      tuple6(1, 2, 3, 4, 5, 6),
    );
  });

  test('Codec.forProduct7', () {
    _codecTest(
      Codec.forProduct7(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        (a, b, c, d, e, f, g) => tuple7(a, b, c, d, e, f, g),
        id,
      ),
      tuple7(1, 2, 3, 4, 5, 6, 7),
    );
  });

  test('Codec.forProduct8', () {
    _codecTest(
      Codec.forProduct8(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        (a, b, c, d, e, f, g, h) => tuple8(a, b, c, d, e, f, g, h),
        id,
      ),
      tuple8(1, 2, 3, 4, 5, 6, 7, 8),
    );
  });

  test('Codec.forProduct9', () {
    _codecTest(
      Codec.forProduct9(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        (a, b, c, d, e, f, g, h, i) => tuple9(a, b, c, d, e, f, g, h, i),
        id,
      ),
      tuple9(1, 2, 3, 4, 5, 6, 7, 8, 9),
    );
  });

  test('Codec.forProduct9', () {
    _codecTest(
      Codec.forProduct9(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        (a, b, c, d, e, f, g, h, i) => tuple9(a, b, c, d, e, f, g, h, i),
        id,
      ),
      tuple9(1, 2, 3, 4, 5, 6, 7, 8, 9),
    );
  });

  test('Codec.forProduct10', () {
    _codecTest(
      Codec.forProduct10(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        (a, b, c, d, e, f, g, h, i, j) => tuple10(a, b, c, d, e, f, g, h, i, j),
        id,
      ),
      tuple10(1, 2, 3, 4, 5, 6, 7, 8, 9, 10),
    );
  });

  test('Codec.forProduct11', () {
    _codecTest(
      Codec.forProduct11(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        (a, b, c, d, e, f, g, h, i, j, k) =>
            tuple11(a, b, c, d, e, f, g, h, i, j, k),
        id,
      ),
      tuple11(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11),
    );
  });

  test('Codec.forProduct12', () {
    _codecTest(
      Codec.forProduct12(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l) =>
            tuple12(a, b, c, d, e, f, g, h, i, j, k, l),
        id,
      ),
      tuple12(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12),
    );
  });

  test('Codec.forProduct13', () {
    _codecTest(
      Codec.forProduct13(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m) =>
            tuple13(a, b, c, d, e, f, g, h, i, j, k, l, m),
        id,
      ),
      tuple13(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13),
    );
  });

  test('Codec.forProduct14', () {
    _codecTest(
      Codec.forProduct14(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        'n'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m, n) =>
            tuple14(a, b, c, d, e, f, g, h, i, j, k, l, m, n),
        id,
      ),
      tuple14(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14),
    );
  });

  test('Codec.forProduct15', () {
    _codecTest(
      Codec.forProduct15(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        'n'.integer,
        'o'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m, n, o) =>
            tuple15(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o),
        id,
      ),
      tuple15(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15),
    );
  });

  test('Codec.forProduct16', () {
    _codecTest(
      Codec.forProduct16(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        'n'.integer,
        'o'.integer,
        'p'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p) =>
            tuple16(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p),
        id,
      ),
      tuple16(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16),
    );
  });

  test('Codec.forProduct17', () {
    _codecTest(
      Codec.forProduct17(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        'n'.integer,
        'o'.integer,
        'p'.integer,
        'q'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q) =>
            tuple17(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q),
        id,
      ),
      tuple17(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17),
    );
  });

  test('Codec.forProduct18', () {
    _codecTest(
      Codec.forProduct18(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        'n'.integer,
        'o'.integer,
        'p'.integer,
        'q'.integer,
        'r'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r) =>
            tuple18(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r),
        id,
      ),
      tuple18(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18),
    );
  });

  test('Codec.forProduct19', () {
    _codecTest(
      Codec.forProduct19(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        'n'.integer,
        'o'.integer,
        'p'.integer,
        'q'.integer,
        'r'.integer,
        's'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s) =>
            tuple19(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s),
        id,
      ),
      tuple19(
          1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19),
    );
  });

  test('Codec.forProduct20', () {
    _codecTest(
      Codec.forProduct20(
        'a'.integer,
        'b'.integer,
        'c'.integer,
        'd'.integer,
        'e'.integer,
        'f'.integer,
        'g'.integer,
        'h'.integer,
        'i'.integer,
        'j'.integer,
        'k'.integer,
        'l'.integer,
        'm'.integer,
        'n'.integer,
        'o'.integer,
        'p'.integer,
        'q'.integer,
        'r'.integer,
        's'.integer,
        't'.integer,
        (a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t) =>
            tuple20(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t),
        id,
      ),
      tuple20(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
          20),
    );
  });
}

void _codecTest<A>(Codec<A> codec, A value) {
  codec.decode(codec.encode(value)).fold(
        (err) => fail('${codec.runtimeType} failed: $err'),
        (decoded) => expect(decoded, value),
      );
}


test/error_test.dart
import 'package:pj/pj.dart';
import 'package:test/test.dart';

void main() {
  test('DecodingError.withReason', () {
    expect(DecodingError.apply('foo').withReason('bar').reason, 'bar');
    expect(DecodingError.missingField('foo').withReason('bar').reason, 'bar');
    expect(DecodingError.parsingFailure('foo').withReason('bar').reason, 'bar');
  });

  test('DecodingError.hashcode equality', () {
    expect(
      DecodingError.missingField('1234567890').hashCode,
      DecodingError.missingField('1234567890').hashCode,
    );
  });

  test('DecodingError.toString() equality', () {
    expect(
      const DecodingError('i messed up.').toString(),
      const DecodingError('i messed up.').toString(),
    );
  });

  test('MissingField.toString() equality', () {
    expect(
      DecodingError.missingField('Miss u.').toString(),
      DecodingError.missingField('Miss u.').toString(),
    );
  });

  test('ParsingError.toString() equality', () {
    expect(
      DecodingError.parsingFailure('Parsnippity!').toString(),
      DecodingError.parsingFailure('Parsnippity!').toString(),
    );
  });
}


lib/syntax.dart
import 'package:dartz/dartz.dart';
import 'package:pj/src/codec.dart';

extension CodecStringOps on String {
  Codec<BigInt> get bigInt => Codec.bigint.at(this);

  Codec<bool> get boolean => Codec.boolean.at(this);

  Codec<DateTime> get dateTime => Codec.dateTime.at(this);

  Codec<double> get dubble => Codec.dubble.at(this);

  Codec<Duration> get duration => Codec.duration.at(this);

  Codec<IList<A>> ilistOf<A>(Codec<A> elementCodec) =>
      Codec.ilistOf(elementCodec).at(this);

  Codec<int> get integer => Codec.integer.at(this);

  Codec<List<A>> listOf<A>(Codec<A> elementCodec) =>
      Codec.listOf(elementCodec).at(this);

  Codec<num> get number => Codec.number.at(this);

  Codec<A> of<A>(Codec<A> codec) => codec.at(this);

  Codec<Map<String, dynamic>> get object => Codec.object.at(this);

  Codec<String> get string => Codec.string.at(this);
}

Codec<BigInt> Function(String s) get bigInt => (key) => key.bigInt;

Codec<bool> Function(String s) get boolean => (key) => key.boolean;

Codec<DateTime> Function(String s) get dateTime => (key) => key.dateTime;

Codec<double> Function(String s) get dubble => (key) => key.dubble;

Codec<Duration> Function(String s) get duration => (key) => key.duration;

Codec<IList<A>> Function(String s) ilistOf<A>(Codec<A> elementCodec) =>
    (key) => key.ilistOf(elementCodec);

Codec<int> Function(String s) get integer => (key) => key.integer;

Codec<List<A>> Function(String s) listOf<A>(Codec<A> elementCodec) =>
    (key) => key.listOf(elementCodec);

Codec<num> Function(String s) get number => (key) => key.number;

Codec<Map<String, dynamic>> Function(String s) get object =>
    (key) => key.object;

Codec<String> Function(String s) get string => (key) => key.string;


lib/src/error.dart
class DecodingError {
  final String reason;

  const DecodingError(this.reason);

  static DecodingError apply(String reason) => DecodingError(reason);

  static DecodingError missingField(String key) => MissingFieldFailure._(key);

  static DecodingError parsingFailure(String reason) =>
      ParsingFailure._(reason);

  DecodingError withReason(String reason) => DecodingError(reason);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is DecodingError &&
          runtimeType == other.runtimeType &&
          reason == other.reason;

  @override
  int get hashCode => runtimeType.hashCode ^ reason.hashCode;

  @override
  String toString() => 'DecodingError: $reason';
}

class MissingFieldFailure extends DecodingError {
  const MissingFieldFailure._(String reason) : super(reason);

  @override
  DecodingError withReason(String reason) => MissingFieldFailure._(reason);

  @override
  String toString() => 'MissingFieldFailure: $reason';
}

class ParsingFailure extends DecodingError {
  const ParsingFailure._(String reason) : super(reason);

  @override
  DecodingError withReason(String reason) => ParsingFailure._(reason);

  @override
  String toString() => 'ParsingFailure: $reason';
}


example/main.dart
import 'package:dartz/dartz.dart';
import 'package:pj/pj.dart';
import 'package:pj/syntax.dart';

// ignore_for_file: avoid_print

void main() {
  final people = [
    Person('Albert', 'Einstein', DateTime(1879, 3, 14), 7, some(true),
        [Pet('Physics', none(), 12)]),
    Person('Douglas', 'Adams', DateTime(1952, 5, 11), 42, none(), []),
    Person('Dolly', 'Parton', DateTime(1946, 1, 19), 95, some(false), []),
  ];

  // Encode and attempt decode of Albert Einstein
  final albertJson = Person.codec.encode(people[0]);

  Person.codec.decode(albertJson).fold(
        (err) => print('[✘] Failed to decode Einstein: $err'),
        (albert) => print(
          '[✔] Alberts lucky number is ${albert.luckyNumber} and he has ${albert.pets.length} pet(s).',
        ),
      );

  // Encode the entire list of people
  final peopleJson = Codec.listOf(Person.codec).encode(people);

  // Decode the entire list of people
  Codec.listOf(Person.codec).decode(peopleJson).fold(
        (err) => print('[✘] Nobody is coming to the party: $err'),
        (inviteList) => print(
            '[✔] ${inviteList.map((p) => p.firstName).join(", ")} will be attending the party.'),
      );

  final petJson = {'pet-name': 'Ribs', 'age': 13};

  Pet.codecAlt.decode(petJson).fold(
        (err) => print('[✘] Parse failed: $err'),
        (ribs) => print(
            '[✔] ${ribs.name} is the best boy. Give him ${ribs.bellyRubs} belly rubs!'),
      );

  // Decoding pet json with a Person decoder will fail with an error
  Person.codec.decode(petJson).fold(
        (err) =>
            print("[✔] Of course you can't turn a pet into a person: $err"),
        (humanPet) => print(
          "[✘] Your scientists were so preoccupied with whether or not they could, they didn't stop to think if they should.",
        ),
      );
}

class Person {
  final String firstName;
  final String? lastName;
  final DateTime birthday;
  final int luckyNumber;
  final Option<bool> registeredVoter;
  final List<Pet> pets;

  Person(
    this.firstName,
    this.lastName,
    this.birthday,
    this.luckyNumber,
    this.registeredVoter,
    this.pets,
  );

  /// Our [Person] codec that will serialize a value to JSON object
  static final codec = Codec.forProduct6(
    // Provide the individual key-value codecs
    'firstName'.string.ensure((s) => s.isNotEmpty, 'Missing first name.'),
    'lastName'.string.nullable,
    'birthday'.dateTime,
    'luckyNumber'.integer,
    'registeredVoter'.boolean.optional,
    'pets'.listOf(Pet.codecAlt),
    // Provide how to turn the individual values into out Product type (Pet)
    Person.new,
    // Provide how to turn our product type into a tuple
    (person) => tuple6(person.firstName, person.lastName, person.birthday,
        person.luckyNumber, person.registeredVoter, person.pets),
  );

  /// Alternative codec definition equivalent to the one above.
  static final codecAlt = Codec.forProduct6(
    string('firstName').ensure((s) => s.isNotEmpty, 'Missing first name'),
    string('lastName').nullable,
    dateTime('birthday'),
    integer('luckyNumber'),
    boolean('registeredVoter').optional,
    listOf(Pet.codec)('pets'),
    Person.new,
    (person) => tuple6(person.firstName, person.lastName, person.birthday,
        person.luckyNumber, person.registeredVoter, person.pets),
  );
}

class Pet {
  final String name;
  final Option<int> age;
  final int bellyRubs;

  Pet(this.name, this.age, this.bellyRubs);

  /// Our [Pet] codec that will serialize a value to JSON object
  static final codec = Codec.forProduct3(
    'pet-name'.string,
    'age'.integer.optional,
    'belly-rubs'.integer.withDefault(8675309),
    Pet.new,
    (pet) => tuple3(pet.name, pet.age, pet.bellyRubs),
  );

  /// Alternative way to define our [Pet] codec, equivalent to the
  /// definition above.
  static final codecAlt = Codec.forProduct3(
    string('pet-name'),
    integer('age').optional,
    integer('belly-rubs').withDefault(8675309),
    Pet.new,
    (pet) => tuple3(pet.name, pet.age, pet.bellyRubs),
  );

  /// Another alternative, lower level, way to define our [Pet] codec
  /// This method opens a few addional doors that give you more power
  /// but also more ways to potentially shoot yourself in the foot.
  static final codecLowLevel = Codec.forProduct3(
    Codec.string.at('pet-name'),
    Codec.integer.at('age').optional,
    Codec.integer.at('belly-rubs').withDefault(8675309),
    Pet.new,
    (pet) => tuple3(pet.name, pet.age, pet.bellyRubs),
  );
}
